src/base/scene.cu

```diff
@@ -0,0 +1,216 @@
+#include "object.h"
+#include "scene.h"
+
+#define TINYGLTF_IMPLEMENTATION
+#define STB_IMAGE_IMPLEMENTATION
+#include "tiny_gltf.h"
+
+__host__ gm::Scene::Scene(std::string filepath) {
+  tinygltf::Model model;
+  tinygltf::TinyGLTF loader;
+  std::string err;
+  std::string warn;
+
+  bool ret = false;
+  if (filepath.substr(filepath.length() - 4, filepath.length()) == ".glb") {
+    ret = loader.LoadBinaryFromFile(&model, &err, &warn, filepath);
+  } else if (filepath.substr(filepath.length() - 5, filepath.length()) ==
+             ".gltf") {
+    ret = loader.LoadASCIIFromFile(&model, &err, &warn, filepath);
+  } else {
+    // Error here because we got an unknown file extension.
+    // TODO error handling
+  }
+
+  if (!warn.empty()) {
+    printf("Warn: %s\n", warn.c_str());
+  }
+
+  if (!err.empty()) {
+    printf("Err: %s\n", err.c_str());
+  }
+
+  if (!ret) {
+    printf("Failed to parse glTF\n");
+  }
+
+  tinygltf::Scene gltfScene = model.scenes[model.defaultScene];
+  std::cout << "Loading scene " << gltfScene.name << std::endl;
+  this->objects = load_objects(gltfScene.nodes, model);
+
+  for (size_t i = 0; i < this->objects.size(); ++i) {
+    if (this->objects[i] != nullptr) {
+      std::cout << this->objects[i]->name
+                << " location:" << this->objects[i]->location[0] << ", "
+                << this->objects[i]->location[1] << ", "
+                << this->objects[i]->location[2] << std::endl;
+    }
+  }
+}
+
+std::vector<std::shared_ptr<gm::Object>> gm::Scene::load_objects(
+    std::vector<int> node_ids, tinygltf::Model model,
+    std::shared_ptr<Object> parent) {
+  std::vector<std::shared_ptr<Object>> objects;
+  std::shared_ptr<Object> current;
+  std::cout << "Loading " << node_ids.size() << " objects" << std::endl;
+  for (size_t i = 0; i < node_ids.size(); ++i) {
+    tinygltf::Node node = model.nodes[node_ids[i]];
+    if (node.mesh != -1) {
+      current = load_mesh(node, model);
+    } else if (node.camera != -1) {
+      current = load_camera(node, model);
+    } else {
+      current = load_empty(node, model);
+    }
+    current->parent = parent;
+    current->children = load_objects(node.children, model, current);
+    objects.push_back(current);
```

Comment:
Instead of using push_back() can we pre-allocate the scene object buffer based on the data parsed through tiny_gltf_loader? I can this bottlenecking the load times later on when we load more complex scenes and the vector has to be resized frequently.

Reply:
Reserved memory beforehand (7c7e1d2)